mev_inspect/arbitrages.py

```diff
@@ -23,70 +23,124 @@ def get_arbitrages(swaps: List[Swap]) -> List[Arbitrage]:
-            current_swap.token_out_address,
-            other_swaps,
+        valid_routes = filter(
+            lambda route: _swaps_have_valid_order_wrt_trace_address(route), routes
         )
-
-        if next_swap is None:
-            return None
-
-        swap_path.append(next_swap)
-        current_swap = next_swap
-
-        if (
-            current_swap.to_address == start_swap.from_address
-            and current_swap.token_out_address == start_swap.token_in_address
-        ):
-
-            start_amount = start_swap.token_in_amount
-            end_amount = current_swap.token_out_amount
+        for route in valid_routes:
+            start_amount = route[0].token_in_amount
+            end_amount = route[-1].token_out_amount
             profit_amount = end_amount - start_amount
 
-            return Arbitrage(
-                swaps=swap_path,
-                block_number=start_swap.block_number,
-                transaction_hash=start_swap.transaction_hash,
-                account_address=start_swap.from_address,
-                profit_token_address=start_swap.token_in_address,
+            arb = Arbitrage(
+                swaps=route,
+                block_number=route[0].block_number,
+                transaction_hash=route[0].transaction_hash,
+                account_address=route[0].from_address,
+                profit_token_address=route[0].token_in_address,
                 start_amount=start_amount,
                 end_amount=end_amount,
                 profit_amount=profit_amount,
             )
+            all_arbitrages.append(arb)
 
-    return None
-
+    return all_arbitrages
 
-def _get_swap_from_address(
-    address: str, token_address: str, swaps: List[Swap]
-) -> Optional[Swap]:
-    for swap in swaps:
-        if swap.pool_address == address and swap.token_in_address == token_address:
-            return swap
 
-    return None
+def _get_all_start_end_swaps(swaps: List[Swap]) -> List[Tuple[Swap, Swap]]:
+    """
+    Gets the set of all possible opening and closing swap pairs in an arbitrage via
+    - swap[start].token_in == swap[end].token_out
+    - swap[start].token_in_amount < swap[end].token_out_amount (assumes profitable)
+    """
+    # TODO: Consider also filtering by start.from == end.to -- duplicate for now.
+    results: List[Tuple[Swap, Swap]] = []
+    for potential_start_swap in swaps:
+        for potential_end_swap in swaps:
+
+            # Checking by profitability can over-discover due to MEV-bot dust. Ex: 3 pool arb, second tx does not use all of output of first, we'd classify two start/end points. These have to be filtered separately.
+            if (
+                potential_start_swap.token_in_address
+                == potential_end_swap.token_out_address
+                and potential_start_swap.token_in_amount
+                < potential_end_swap.token_out_amount
+            ):
+                results.append((potential_start_swap, potential_end_swap))
```

Comment:
is there a way we could keep non-profitable arbs? - it can be neat to see that people that ended up executing something that net lost them money

for the case where someone doesn't use all of the output in an intermediate step, why does that result in two start / ends?

Reply:
Notes on why relying on profitability is necessary in comment#<FILL THIS IN SAM>.

On not using all the output-
If you have `A->B`/`B->C`/`C->A` and you're trying to find a start/end based on profitability and you have
- `(A->B).a_in < (C->A).a_out` - due to successful arbitrage
- `(B->C).b_in < (A->B).b_out` - due to dust / rounding error
There is no way to tell which is the valid start and end. Of course you could add a config that says drop arbs below a certain amount of profitability, but that's its own can of worms.